### PR TITLE
fix: add minimimal emscripten support via non-HTTP sources

### PR DESCRIPTION
**Pull request by agoose77**
#868 fixed some import errors, but we still used threads for non-HTTP sources. In this PR, we close the gap and make it possible to disable threads for all sources. By default, this is done for emscripten platforms.

I didn't add HTTP support here — it's possible with e.g. `XMLHttpRequest`, but we'll need to spend some more cycles on it. Crucially, CORS means that this can be fiddly, but it would be more performant for emscripten platforms where `mmap` is not supported.

See https://agoose77.github.io/pyhep-23-jupyterlite/lab/index.html for a repo that uses this patch.

### CODE REVIEW

**Comment by ioanaif** [comment]
Nice! 
For me the file opening part is still not working in the demo. However, I am not sure if it is a problem caused by my browser (firefox - happened in the past for CoDaS)

**Comment by agoose77** [comment]
@ioanaif I use Firefox too! Does it give a useful traceback e.g. `FileNotFoundError`? If so, it might help to force-refresh the page.

**Comment by ioanaif** [comment]
Yes. Last time this happened for my browser because the file was too large

File: src/uproot/source/http.py [hunk 1]
```diff
@@ -298,25 +299,24 @@ def multifuture(source, ranges, futures, results):
         )
 
         def task(resource):
+            nonlocal connection
```
Comment (jpivarski) [inline src/uproot/source/http.py:302]: That's right: `connection[0]` was a Python2ism. `global` is not the right scope, and `nonlocal` didn't exist.

**Review by jpivarski — APPROVED** [review]
The new argument, `no_threads`, overrides `num_*workers` and there should be some note in the documentation to indicate this. (It's always a little more confusing when one argument changes what another does.)

Here are all of the places where `num_*workers` arguments are defined:

```bash
% fgrep -r 'workers (int' src/uproot/ --include='*.py'
src/uproot/reading.py:    * num_workers (int; 1)
src/uproot/reading.py:    * num_fallback_workers (int; 10)
src/uproot/reading.py:    * num_workers (int; 1)
src/uproot/reading.py:    * num_fallback_workers (int; 10)
src/uproot/_dask.py:    * num_workers (int; 1)
src/uproot/_dask.py:    * num_fallback_workers (int; 10)
src/uproot/behaviors/TBranch.py:    * num_workers (int; 1)
src/uproot/behaviors/TBranch.py:    * num_fallback_workers (int; 10)
src/uproot/behaviors/TBranch.py:    * num_workers (int; 1)
src/uproot/behaviors/TBranch.py:    * num_fallback_workers (int; 10)
```

Actually, I don't think `no_threads` is documented at all, including the fact that its default depends on `sys.platform`.

File: src/uproot/source/http.py (hunk 1 above)
Comment (agoose77) [reply src/uproot/source/http.py:302]: Hmm, you're right. At one point I was making this an implicit thing, but there's sense in exposing it.

**Comment by jpivarski** [comment]
> > Actually, I don't think no_threads is documented at all, including the fact that its default depends on sys.platform.
> 
> Hmm, you're right. At one point I was making this an implicit thing, but there's sense in exposing it.

If `no_threads` is in the list of `**options`, then it's already exposed. A user could override that manually.

**Comment by agoose77** [comment]
Let me hold this until I finish some last changes :)

**Comment by agoose77** [comment]
@ioanaif and @jpivarski, what do you think about testing this PR? We could just parameterise the `use_threads` argument, but that will more than double our test time.

**Comment by jpivarski** [comment]
We could double our test time on a subset of tests. Pick one that seems to be focusing on backend features, one that tests HTTP and XRootD with various options, non-existent files, etc., and double those with `use_threads=True/False`.

**Comment by agoose77** [comment]
I've updated the tests. The documentation is probably not perfect; we talk about the executor being a threaded executor, but sometimes it's a trivial executor. I'm happy to not worry about that for now, as I think we should drop our executor abstraction in the near future in favour of the existing solution in Python 3's `concurrent.futures.`

File: src/uproot/_dask.py [hunk 2]
```diff
@@ -127,6 +127,7 @@ def dask(
     * timeout (float for HTTP, int for XRootD; 30)
     * max_num_elements (None or int; None)
     * num_workers (int; 1)
+    * use_threads (bool; False if emscripten, else True)
```
Comment (jpivarski) [inline src/uproot/_dask.py:130]:
```suggestion
    * use_threads (bool; False on the emscripten platform (i.e. in a web browser), else True)
```

File: src/uproot/behaviors/TBranch.py [hunk 3]
```diff
@@ -160,6 +160,7 @@ def iterate(
     * timeout (float for HTTP, int for XRootD; 30)
     * max_num_elements (None or int; None)
     * num_workers (int; 1)
+    * use_threads (bool; False if emscripten, else True)
```
Comment (jpivarski) [inline src/uproot/behaviors/TBranch.py:163]:
```suggestion
    * use_threads (bool; False on the emscripten platform (i.e. in a web browser), else True)
```

File: src/uproot/behaviors/TBranch.py [hunk 4]
```diff
@@ -332,6 +333,7 @@ def concatenate(
...
```
Comment (jpivarski) [inline src/uproot/behaviors/TBranch.py:336]:
```suggestion
    * use_threads (bool; False on the emscripten platform (i.e. in a web browser), else True)
```

File: src/uproot/reading.py [hunk 5]
```diff
@@ -85,6 +85,7 @@ def open(
     * timeout (float for HTTP, int for XRootD; 30)
     * max_num_elements (None or int; None)
     * num_workers (int; 1)
+    * use_threads (bool; False if emscripten, else True)
```
Comment (jpivarski) [inline src/uproot/reading.py:88]:
```suggestion
    * use_threads (bool; False on the emscripten platform (i.e. in a web browser), else True)
```

File: src/uproot/reading.py [hunk 6]
```diff
@@ -541,6 +542,7 @@ class ReadOnlyFile(CommonFileMethods):
...
```
Comment (jpivarski) [inline src/uproot/reading.py:545]:
```suggestion
    * use_threads (bool; False on the emscripten platform (i.e. in a web browser), else True)
```

**Review by jpivarski — APPROVED** [review]
I'm only making a documentation suggestion (for users who haven't heard of emscripten).

**Pull request merged**